src/sources/aws_s3/sqs.rs

```diff
@@ -0,0 +1,517 @@
+            emit!(SqsS3EventRecordIgnoredInvalidEvent {
+                bucket: &s3_event.s3.bucket.name,
+                key: &s3_event.s3.object.key,
+                kind: &s3_event.event_name.kind,
+                name: &s3_event.event_name.name,
+            });
+            return Ok(());
+        }
+
+        // S3 has to send notifications to a queue in the same region so I don't think this will
+        // actually ever be it unless messages are being forwarded from one queue to another
+        if self.region.name() != s3_event.aws_region {
+            return Err(ProcessingError::WrongRegion {
+                bucket: s3_event.s3.bucket.name.clone(),
+                key: s3_event.s3.object.key.clone(),
+                region: s3_event.aws_region,
+            });
+        }
+
+        let object = self
+            .s3_client
+            .get_object(GetObjectRequest {
+                bucket: s3_event.s3.bucket.name.clone(),
+                key: s3_event.s3.object.key.clone(),
+                ..Default::default()
+            })
+            .await
+            .context(GetObject {
+                bucket: s3_event.s3.bucket.name.clone(),
+                key: s3_event.s3.object.key.clone(),
+            })?;
+
+        let metadata = object.metadata;
+        let timestamp = object
+            .last_modified
+            .and_then(|t| {
+                DateTime::parse_from_rfc2822(&t)
+                    .map(|ts| Utc.timestamp(ts.timestamp(), ts.timestamp_subsec_nanos()))
+                    .ok()
+            })
+            .unwrap_or_else(Utc::now);
+
+        match object.body {
+            Some(body) => {
+                let r = super::s3_object_decoder(
+                    self.compression,
+                    &s3_event.s3.object.key,
+                    object.content_encoding.as_deref(),
+                    object.content_type.as_deref(),
+                    body,
+                );
+
+                // Record the read error saw to propagate up later so we avoid ack'ing the SQS
+                // message
+                //
+                // String is used as we cannot take clone std::io::Error to take ownership in
+                // closure
+                //
+                // FramedRead likely stops when it gets an i/o error but I found it more clear to
+                // show that we `take_while` there hasn't been an error
+                //
+                // This can result in objects being partially processed before an error, but we
+                // prefer duplicate lines over message loss. Future work could include recording
+                // the offset of the object that has been read, but this would only be relevant in
+                // the case that the same vector instance processes the same message.
+                let mut read_error: Option<String> = None;
+                let mut lines: Box<dyn Stream<Item = Bytes> + Send + Unpin> = Box::new(
```

Comment:
What about having `Item = Result<Bytes, io::Error>` and handle the errors at a later step?

Reply:
I actually tried something like this, but `LineAgg` needs `Bytes` so I'd just need to catch it there, which isn't much later. I'm definitely interested in improving this though as it seems like the pattern probably won't be uncommon.